DOC: Point Contributing page to new NEP 45

bjnath opened the pull request:
Moved the link for C style in the "Contributing to NumPy" page to NEP 45.

In NEP 45, replaced explicit link to PEP 7 with a :pep: link.

Fixes problem raised in https://github.com/numpy/numpy/pull/16430#issuecomment-638789051.

The new link points to the NEP's URL. Since NEPs are built separately from the rest of the docs, I wasn't sure an intersphinx link would work. I searched the docs for markup like `:nep:` or `:ref:.*nep` or `:doc:.*nep` but found nothing.

bjnath committed:
DOC: Point Contributing page to new NEP 45

Moved the link for C style to NEP 45.

In NEP 45, replaced explicit link to PEP 7 with a :pep: link.

mattip commented:
In order to add neps to intersphinx you need to add the link https://numpy.org/neps/objects.inv to it to the list in [conf.py](https://github.com/numpy/numpy/blob/v1.19.0rc2/doc/source/conf.py#L242)

mattip commented:
This would unlock changing the direct links in `doc/HOWTO_RELEASE.rst.txt`, `doc/source/dev/howto-docs.rst`, `numpy/doc/dispatch.py`, and `numpy/lib/format.py` (found via `git grep "numpy.org/neps/nep" |grep -v "^doc/neps" |grep -v "^doc/source/release" |grep -v "^numpy/lib/financial"`

bjnath commented:
Why were the `grep -v` dirs excluded?

mattip commented:
- `neps` goes to a separate doc build, if we want to refactor all the internal links that should be a separate PR
- `release` are all the release notes, in principle we do not touch them; too much churn
- `financial` is to be deprecated soon, no need to touch it.

rossbar commented:
> In order to add neps to intersphinx you need to add the link https://numpy.org/neps/objects.inv to it to the list in conf.py

#16502 contains a proposal to add the NEPs to the intersphinx mapping. Thanks @bjnath and @mattip for the helpful discussion.

bjnath commented:
I need a bit of markup help. I have a case where the intersphinx link needs to be in a link block:
```
.. _NEP 29 : https://numpy.org/neps/nep-0029-deprecation_policy.html
```
When I try
```
.. _NEP 29 : :doc:`neps:nep-0029-deprecation_policy`
```
Sphinx sees it only as literal text. 

Inline cases, like
```
:doc:`NEP 29<neps:nep-0029-deprecation_policy>`
```
work fine.

[Sphinx doc](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html) says

> Basically, you only need to write :role:`target`, and a link will be created to the item named target of the type indicated by role. The link’s text will be the same as target.

 Basically, I'm thinking they forgot something.

Does anyone know if role+intersphinx works in a link block? 

And if it doesn't, which would we rather have:

* Two inline references to NEP 29 that use :neps:
* A link-block reference with a URL

mattip commented:
Two inline references to NEP 29 that use :neps: is fine.

bjnath committed:
DOC: Use intersphinx for NEP references

Per #16500 discussion, adds link updates to PR.
Replaces NEP URLs with intersphinx links in these
files:

* doc/HOWTO_RELEASE.rst.txt
* doc/source/dev/howto-docs.rst
* numpy/doc/dispatch.py
* numpy/lib/format.py

and incorporates @rossbar's #16502 change of `config.py`.

bjnath commented:
Per @mattip's suggestion, 7764b5f replaces the NEP URL's with intersphinx links in these files:

* doc/HOWTO_RELEASE.rst.txt
* doc/source/dev/howto-docs.rst
* numpy/doc/dispatch.py
* numpy/lib/format.py

and incorporates @rossbar's change to `config.py`.

rossbar commented on doc/HOWTO_RELEASE.rst.txt:
```diff
@@ -40,16 +40,17 @@ Release Scripts
-2020, this will be Python >= 3.6. We test NumPy against all these versions
-every time we merge code to master.  Binary installers may be available for a
-subset of these versions (see below).
+:doc:`NEP 29<neps:nep-0029-deprecation_policy>` outlines which Python versions
```
```suggestion
:ref:`nep0029` outlines which Python versions
```
The links/abbreviations can be made shorter with intersphinx I think. This will replace the `:ref:` with the full title of the document.

rossbar commented on doc/HOWTO_RELEASE.rst.txt:
```diff
@@ -40,16 +40,17 @@ Release Scripts
-We build binary wheels for OSX that are compatible with Python.org Python,
-system Python, homebrew and macports - see this `OSX wheel building summary
+OS X versions >= 10.9 are supported, for Python version support see
+:doc:`NEP 29<neps:nep-0029-deprecation_policy>`. We build binary wheels for
```
```suggestion
:ref:`NEP 29 <nep0029>`. We build binary wheels for
```
... and this would use the shorter name `NEP 29` for the link.

rossbar commented on doc/source/dev/howto-docs.rst:
```diff
@@ -35,8 +35,9 @@ Current vision for the documentation: NEP 44
 Recently, the NumPy community approved a *NumPy Enhancement Proposal (NEP)*
-about documentation, `NEP 44 - Restructuring the NumPy Documentation
-<https://www.numpy.org/neps/nep-0044-restructuring-numpy-docs>`__.
+about documentation,
+:doc:`NEP 44 - Restructuring the NumPy Documentation
+     <neps:nep-0044-restructuring-numpy-docs>`.
```
```suggestion
about documentation, :ref:`nep44`.
```

rossbar commented on doc/source/dev/index.rst:
```diff
@@ -189,7 +189,7 @@ Stylistic Guidelines
    import numpy as np
 
-* For C code, see the :ref:`numpy-c-style-guide<style_guide>`
+* For C code, see `NEP 45 <https://numpy.org/neps/nep-0045-c_style_guide.html>`_.
```
If you add `.. _NEP45:` to the top of the nep (see #16450 for example), you could reference it via intersphinx here as well, e.g. `` :ref:`NEP 45 <nep45>` ``

rossbar reviewed (commented):
Looks good! With intersphinx I think there is a shorter syntax for the links: instead of `` :doc:`NEP X <neps:full_nep_document_title>` ``, `` :ref:`NEP XX <nepXX>` `` could be used instead (see specific comments for some examples).

The links are fine the way they are too, but just wanted to point out the shorter option.

bjnath committed:
DOC: Add NEP45 and NEP29 targets, use <NEPxx> links

Per gh-16500, changed link style to :ref:`<NEPxx>`.

Added a NEP45 target to NEP 45, and added a NEP29 target
to NEP 29 since it has the irregular NEP0029 target.

The two new NEP targets aren't picked up, though the
older ones are. Tried make clean in neps and docs.

```
  /home/bjn/numpy_git/numpy-1/doc/source/dev/index.rst:191: WARNING: undefined label: nep45 (if the link has no caption the label must precede a section header)
  HOWTO_RELEASE.rst.txt:43: WARNING: undefined label: nep29 (if the link has no caption the label must precede a section header)
  HOWTO_RELEASE.rst.txt:50: WARNING: undefined label: nep29 (if the link has no caption the label must precede a section header)
```

The nep objects.inv does seem to have both the nep29 and nep44 targets.

bjnath commented:
e9b8bbd makes suggested changes, also adds an NEP29 target to NEP 29 because unlike the others it had leading zeros. 

Although the older NEPs link OK via `:ref:`, the two new targets (NEP45, NEP29) do not; `make html` fails with `undefined label`.

rgommers commented:
> Although the older NEPs link OK via `:ref:`, the two new targets (NEP45, NEP29) do not; `make html` fails with `undefined label`.

Probably because it's two separate doc builds. The added NEP labels are present in this PR, but not yet in intersphinx of the deployed docs.

I'll have a look at how to merge this.

rgommers commented:
Okay I verified locally that this all looks correct. It's not possible to introduce a NEP label and `:ref:` it in a single PR. Rather than requiring a separate PR, I'll just merge this. The NEPs build will be deployed, and subsequent PRs will trigger a rebuild to update master.

Merging, thanks @bjnath. And thanks @rossbar for review.

rgommers commented:
Okay that wasn't the best idea. Should have just reverted the last commit. Did a manual deploy now, that weirdly misses the theme. Will be fixed on next deploy.

rgommers commented:
All fixed now.

bjnath commented:
Sorry to have caused extra effort -- thanks for getting it to work. I hoped dependencies would be satisfied by make clean, build neps, build doc, but evidently it's not that simple.
